ext/zend_test/observer.c

```diff
@@ -69,6 +69,9 @@ static void observer_show_opcode(zend_execute_data *execute_data)
 
 static void observer_begin(zend_execute_data *execute_data)
 {
+	ZEND_ASSERT(!ZEND_USER_CODE(EX(func)->type) ||
+		(EX(opline) >= EX(func)->op_array.opcodes && EX(opline) < EX(func)->op_array.opcodes + EX(func)->op_array.last));
+
```

Comment:
I would recommend to insert this kind of assertions into all other handlers.

Reply:
``make test TESTS="--asan -d zend_test.observer.enabled=1 -d zend_test.observer.observe_all=1 -d zend_test.observer.show_output=0 -d opcache.jit=0 -j16 Zend tests ext/opcache ext/zend_test"`` shows 71 failures. 

Some of them are false positives, but many failures are caused by this assertion. They are mostly related to generators and fibers. This is even without JIT!

Reply:
Added to end handler too. The assertion itself was missing EG(exception_op) handling, good catch.